pandas/_libs/index.pyx

```diff
@@ -294,14 +294,23 @@ cdef class IndexEngine:
         result = np.empty(n_alloc, dtype=np.int64)
         missing = np.empty(n_t, dtype=np.int64)
 
-        # form the set of the results (like ismember)
-        members = np.empty(n, dtype=np.uint8)
-        for i in range(n):
-            val = values[i]
-            if val in stargets:
-                if val not in d:
-                    d[val] = []
-                d[val].append(i)
+        # map each starget to its position in the index
```

Comment:
what if you drop the ``len(stargets) < 5`` and just use it if its monotonic_increasing? does the small case actually make any difference here?

Reply:
If you drop the `len(stargets) < 5`, then we'd be running a binary search against the index for each item in a potentially large set of targets -- runtime should be `O(m log n)` where `m` is the number of items in the set and `n` is the length of the index. Presumably, this would be slower when `m` is large enough compared to the current behavior which is to run through each item in the index and check if it is in the set of targets, which should be `O(n)` assuming constant time checks for whether an item is in the set of targets. Thoughts?

Reply:
can you add a case where this is true in the asv's and compare?

Reply:
yes, in each of the following asv's `stargets` is either length one or length two

> ```
>      [ab9dbd64]       [b704c5bb]
>      <master>         <loc-enhancements>
> -         383±4ms          211±3ms     0.55  indexing.NonNumericSeriesIndexing.time_getitem_list_like('datetime', 'nonunique_monotonic_inc')
> -        59.0±2ms         11.9±1ms     0.20  indexing.CategoricalIndexIndexing.time_get_indexer_list('monotonic_incr')
> -      69.4±0.6ms          445±3μs     0.01  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
> -      66.3±0.3ms          423±1μs     0.01  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
> -      66.1±0.6ms          320±2μs     0.00  indexing.NumericSeriesIndexing.time_ix_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
> -      69.2±0.4ms          330±3μs     0.00  indexing.NumericSeriesIndexing.time_ix_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
> -      65.7±0.3ms          286±3μs     0.00  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
> -      69.3±0.5ms          295±2μs     0.00  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
> 
> SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
> ```